src/Analyzers/Core/CodeFixes/RemoveUnnecessarySuppressions/RemoveUnnecessaryPragmaSuppressionsCodeFixProvider.cs

```diff
@@ -0,0 +1,86 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.RemoveUnnecessarySuppressions
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeFixProviderNames.RemoveUnnecessaryPragmaSuppressions), Shared]
+    internal sealed class RemoveUnnecessaryPragmaSuppressionsCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public RemoveUnnecessaryPragmaSuppressionsCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId);
+
+        internal override CodeFixCategory CodeFixCategory
+            => CodeFixCategory.CodeQuality;
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                // Defensive check that we are operating on the diagnostic on a pragma.
+                if (root.FindTrivia(diagnostic.Location.SourceSpan.Start).HasStructure)
+                {
+                    context.RegisterCodeFix(
+                        new MyCodeAction(c => FixAsync(context.Document, diagnostic, c)),
+                        diagnostic);
+                }
+            }
+        }
+
+        protected override Task FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            using var _ = PooledHashSet<SyntaxNode>.GetInstance(out var seenNodes);
+            foreach (var diagnostic in diagnostics)
+            {
+                RemoveNode(diagnostic.Location, editor, seenNodes);
+
+                foreach (var location in diagnostic.AdditionalLocations)
+                {
+                    RemoveNode(location, editor, seenNodes);
+                }
+            }
+
+            return Task.CompletedTask;
+
+            static void RemoveNode(Location location, SyntaxEditor editor, HashSet<SyntaxNode> seenNodes)
+            {
+                var node = editor.OriginalRoot.FindTrivia(location.SourceSpan.Start).GetStructure()!;
+                if (seenNodes.Add(node))
+                {
+                    editor.RemoveNode(node);
+                }
```

Comment:
you resolved this, but i don't see the documentation.  why do we need this check?

Reply:
Ah, I can add a comment - the reason is we generate a diagnostic at both the pragma disable and matching pragma restore location. Code fix for any one removes the other, so we need to ensure we handle FixAll correctly when both diagnostics are attempted to be fixed.